### ISSUE TITLE
Windows: crawl freezes when multiple schemas are set in DOWNLOAD_HANDLERS

### ISSUE THREAD

**Issue by FuturaTino** [scrapy-plugins/scrapy-playwright#290]
If I have said anything impolite, please forgive me, as I have used a translator. My basic situation is the same as this issue [Spiders hangs](https://github.com/scrapy-plugins/scrapy-playwright/issues/266). I locate the problem to this setting item.
```python
"DOWNLOAD_HANDLERS": {
    "https": "scrapy_playwright.handler.ScrapyPlaywrightDownloadHandler",
    "http": "scrapy_playwright.handler.ScrapyPlaywrightDownloadHandler",
},
```
# Code
When I comment out `DOWNLOAD_HANDLERS`, the spider runs normally. Here is my code, you can directly run it and try.
```python
from pathlib import Path
from scrapy.crawler import CrawlerProcess
import scrapy

class QuotesSpider(scrapy.Spider):
    name = "quotes"

    def start_requests(self):
        urls = [
            "https://quotes.toscrape.com/page/1/",
            "https://quotes.toscrape.com/page/2/",
        ]
        for url in urls:
            yield scrapy.Request(url=url, callback=self.parse,
                                  meta={"playwright": True},)

    def parse(self, response):
        page = response.url.split("/")[-2]
        filename = f"quotes-{page}.html"
        Path(filename).write_bytes(response.body)
        self.log(f"Saved file {filename}")

if __name__=="__main__":
    process = CrawlerProcess(
        settings={
            "TWISTED_REACTOR": "twisted.internet.asyncioreactor.AsyncioSelectorReactor",
            "DOWNLOAD_HANDLERS": {
                "https": "scrapy_playwright.handler.ScrapyPlaywrightDownloadHandler",
                "http": "scrapy_playwright.handler.ScrapyPlaywrightDownloadHandler",
            },
            "CONCURRENT_REQUESTS": 32,
        }
    )
    process.crawl(QuotesSpider)
    process.start()
```
# Result
```plaintext
(env4gpt) D:\Repo\crawl>D:/anaconda3/envs/env4gpt/python.exe d:/Repo/crawl/main.py
2024-07-05 19:36:00 [scrapy.utils.log] INFO: Scrapy 2.11.2 started (bot: scrapybot)
2024-07-05 19:36:00 [scrapy.utils.log] INFO: Versions: lxml 5.2.2.0, libxml2 2.11.7, cssselect 1.2.0, parsel 1.9.1, w3lib 2.2.0, Twisted 24.3.0, Python 3.9.18 (main, Sep 11 2023, 14:09:26) [MSC v.1916 64 bit (AMD64)], pyOpenSSL 24.1.0 (OpenSSL 3.2.2 4 Jun 2024), cryptography 42.0.8, Platform Windows-10-10.0.19045-SP0
2024-07-05 19:36:00 [scrapy.addons] INFO: Enabled addons:
[]
2024-07-05 19:36:00 [py.warnings] WARNING: D:\anaconda3\envs\env4gpt\lib\site-packages\scrapy\utils\request.py:254: ScrapyDeprecationWarning: '2.6' is a deprecated value for the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting.

It is also the default value. In other words, it is normal to get this warning if you have not defined a value for the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting. This is so for backward compatibility reasons, but it will change in a future version of Scrapy.

See the documentation of the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting for information on how to handle this deprecation.
  return cls(crawler)

2024-07-05 19:36:00 [asyncio] DEBUG: Using selector: SelectSelector
2024-07-05 19:36:00 [scrapy.utils.log] DEBUG: Using reactor: twisted.internet.asyncioreactor.AsyncioSelectorReactor
2024-07-05 19:36:00 [scrapy.utils.log] DEBUG: Using asyncio event loop: asyncio.windows_events._WindowsSelectorEventLoop
2024-07-05 19:36:00 [scrapy.extensions.telnet] INFO: Telnet Password: 6b77837564f0d1cc
2024-07-05 19:36:00 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',
 'scrapy.extensions.logstats.LogStats']
2024-07-05 19:36:00 [scrapy.crawler] INFO: Overridden settings:
{'CONCURRENT_REQUESTS': 32,
 'TWISTED_REACTOR': 'twisted.internet.asyncioreactor.AsyncioSelectorReactor'}
2024-07-05 19:36:01 [asyncio] DEBUG: Using proactor: IocpProactor
2024-07-05 19:36:01 [scrapy-playwright] INFO: Started loop on separate thread: <ProactorEventLoop 
running=True closed=False debug=False>
2024-07-05 19:36:01 [asyncio] DEBUG: Using proactor: IocpProactor
2024-07-05 19:36:01 [scrapy-playwright] INFO: Started loop on separate thread: <ProactorEventLoop 
running=True closed=False debug=False>
2024-07-05 19:36:01 [scrapy.middleware] INFO: Enabled downloader middlewares:
['scrapy.downloadermiddlewares.offsite.OffsiteMiddleware',
 'scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware',
 'scrapy.downloadermiddlewares.downloadtimeout.DownloadTimeoutMiddleware',
 'scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware',
 'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware',
 'scrapy.downloadermiddlewares.retry.RetryMiddleware',
 'scrapy.downloadermiddlewares.redirect.MetaRefreshMiddleware',
 'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware',
 'scrapy.downloadermiddlewares.redirect.RedirectMiddleware',
 'scrapy.downloadermiddlewares.cookies.CookiesMiddleware',
 'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware',
 'scrapy.downloadermiddlewares.stats.DownloaderStats']
2024-07-05 19:36:01 [scrapy.middleware] INFO: Enabled spider middlewares:
['scrapy.spidermiddlewares.httperror.HttpErrorMiddleware',
 'scrapy.spidermiddlewares.referer.RefererMiddleware',
 'scrapy.spidermiddlewares.urllength.UrlLengthMiddleware',
 'scrapy.spidermiddlewares.depth.DepthMiddleware']
2024-07-05 19:36:01 [scrapy.middleware] INFO: Enabled item pipelines:
[]
2024-07-05 19:36:01 [scrapy.core.engine] INFO: Spider opened
2024-07-05 19:36:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:36:01 [scrapy.extensions.telnet] INFO: Telnet console listening on 127.0.0.1:6023   
2024-07-05 19:37:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:38:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:39:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:40:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:41:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:42:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:43:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:44:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:45:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:46:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:47:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:48:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:49:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:50:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:51:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-05 19:52:01 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
```
The log continues to repeat "[scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)".
# Conda Environment
```plaintext
openpyxl=3.1.4
scrapy==2.11.2
lxml==5.2.2
python-dotenv==1.0.1
scrapy-playwright==0.0.37
```
# System
Windows 10
# Guess
Is it possible that it is related to the computer system? I currently do not have a computer with a Linux system and cannot verify this. Besides, I am in China, could it be related to the IP address? However, I checked the `ScrapyPlaywrightDownloadHandler` class and did not see any bugs. Thank you for your help.

**Comment by DJousto** [comment]
Same issue for me, on windows 10, same components versions, seems that browser is not launched at all, if I add `PLAYWRIGHT_LAUNCH_OPTIONS = {"headless": False}` I see no browser

**Comment by FuturaTino** [comment]
> Same issue for me, on windows 10, same components versions, seems that browser is not launched at all, if I add `PLAYWRIGHT_LAUNCH_OPTIONS = {"headless": False}` I see no browser

The bug is not fixed yet, you can try `python-playwright` as a alternative solution to develop  your spider program.

**Comment by elacuesta** [comment]
I have been developing the package and testing crawls on Windows 11, installing packages with `pip` in a [virtual environment](https://docs.python.org/3/library/venv.html) (without Conda). I do not currently have access to a Windows 10 system, however the CI Windows build is running on [`windows-latest`](https://github.com/scrapy-plugins/scrapy-playwright/blob/f3b1b25e7ded23ebcc4d5dec73ba0367d0a51d41/.github/workflows/tests.yml#L17-L18): according to the [most recent docs on the images](https://github.com/actions/runner-images/tree/567d4a0ea4a16d7a6e3e03b0f5068b99a7af8111?tab=readme-ov-file#available-images) this is the same as `windows-2022`, which includes [Windows 10.0.20348 Build 2529](https://github.com/actions/runner-images/blob/567d4a0ea4a16d7a6e3e03b0f5068b99a7af8111/images/windows/Windows2022-Readme.md).
There was an earlier issue that involved Conda as well: #244. I could not reproduce it without Conda, given that and the above I'm inclined to believe that Conda might be the issue here.
Please try again without Conda to see if the issue is resolved.

**Comment by FuturaTino** [comment]
> I have been developing the package and testing crawls on Windows 11, installing packages with `pip` in a [virtual environment](https://docs.python.org/3/library/venv.html) (without Conda). I do not currently have access to a Windows 10 system, however the CI Windows build is running on [`windows-latest`](https://github.com/scrapy-plugins/scrapy-playwright/blob/f3b1b25e7ded23ebcc4d5dec73ba0367d0a51d41/.github/workflows/tests.yml#L17-L18): according to the [most recent docs on the images](https://github.com/actions/runner-images/tree/567d4a0ea4a16d7a6e3e03b0f5068b99a7af8111?tab=readme-ov-file#available-images) this is the same as `windows-2022`, which includes [Windows 10.0.20348 Build 2529](https://github.com/actions/runner-images/blob/567d4a0ea4a16d7a6e3e03b0f5068b99a7af8111/images/windows/Windows2022-Readme.md). There was an earlier issue that involved Conda as well: #244. I could not reproduce it without Conda, given that and the above I'm inclined to believe that Conda might be the issue here. Please try again without Conda to see if the issue is resolved.

Thank you for your help. It looks like not the problem of conda in my situation. I closed the conda environment and used the default python.exe on my computer, then downloaded the scrapy-playwright library. Even after ensuring that I'm running the program with the default python.exe rather than the python.exe from the conda environment, the issue persists.

```
D:\Repo\crawl>C:/Python312/python.exe d:/Repo/crawl/main.py
2024-07-09 01:59:19 [scrapy.utils.log] INFO: Scrapy 2.11.2 started (bot: scrapybot)
2024-07-09 01:59:19 [scrapy.utils.log] INFO: Versions: lxml 5.2.2.0, libxml2 2.11.7, cssselect 1.2.0, parsel 1.9.1, w3lib 2.2.1, Twisted 24.3.0, Python 3.12.0 (tags/v3.12.0:0fb18b0, Oct  2 2023, 13:03:39) [MSC v.1935 64 bit (AMD64)], pyOpenSSL 24.1.0 (OpenSSL 3.2.2 4 Jun 2024), cryptography 42.0.8, Platform Windows-10-10.0.19045-SP0
2024-07-09 01:59:19 [scrapy.addons] INFO: Enabled addons:
[]
2024-07-09 01:59:19 [py.warnings] WARNING: C:\Python312\Lib\site-packages\scrapy\utils\request.py:254: ScrapyDeprecationWarning: '2.6' is a deprecated value for the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting.

It is also the default value. In other words, it is normal to get this warning if you have not defined a value for the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting. This is so for backward compatibility reasons, but it will change in 
a future version of Scrapy.

See the documentation of the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting for information on how to handle this deprecation.
  return cls(crawler)

2024-07-09 01:59:19 [asyncio] DEBUG: Using selector: SelectSelector
2024-07-09 01:59:19 [scrapy.utils.log] DEBUG: Using reactor: twisted.internet.asyncioreactor.AsyncioSelectorReactor   
2024-07-09 01:59:19 [scrapy.utils.log] DEBUG: Using asyncio event loop: asyncio.windows_events._WindowsSelectorEventLoop
2024-07-09 01:59:19 [scrapy.extensions.telnet] INFO: Telnet Password: e6ffaca82a959077
2024-07-09 01:59:19 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',
 'scrapy.extensions.logstats.LogStats']
2024-07-09 01:59:19 [scrapy.crawler] INFO: Overridden settings:
{'TWISTED_REACTOR': 'twisted.internet.asyncioreactor.AsyncioSelectorReactor'}
2024-07-09 01:59:19 [asyncio] DEBUG: Using proactor: IocpProactor
2024-07-09 01:59:19 [scrapy-playwright] INFO: Started loop on separate thread: <ProactorEventLoop running=True closed=False debug=False>
2024-07-09 01:59:19 [asyncio] DEBUG: Using proactor: IocpProactor
2024-07-09 01:59:19 [scrapy-playwright] INFO: Started loop on separate thread: <ProactorEventLoop running=True closed=False debug=False>
2024-07-09 01:59:20 [scrapy.middleware] INFO: Enabled downloader middlewares:
['scrapy.downloadermiddlewares.offsite.OffsiteMiddleware',
 'scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware',
 'scrapy.downloadermiddlewares.downloadtimeout.DownloadTimeoutMiddleware',
 'scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware',
 'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware',
 'scrapy.downloadermiddlewares.retry.RetryMiddleware',
 'scrapy.downloadermiddlewares.redirect.MetaRefreshMiddleware',
 'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware',
 'scrapy.downloadermiddlewares.redirect.RedirectMiddleware',
 'scrapy.downloadermiddlewares.cookies.CookiesMiddleware',
 'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware',
 'scrapy.downloadermiddlewares.stats.DownloaderStats']
2024-07-09 01:59:20 [scrapy.middleware] INFO: Enabled spider middlewares:
['scrapy.spidermiddlewares.httperror.HttpErrorMiddleware',
 'scrapy.spidermiddlewares.referer.RefererMiddleware',
 'scrapy.spidermiddlewares.urllength.UrlLengthMiddleware',
 'scrapy.spidermiddlewares.depth.DepthMiddleware']
2024-07-09 01:59:20 [scrapy.middleware] INFO: Enabled item pipelines:
[]
2024-07-09 01:59:20 [scrapy.core.engine] INFO: Spider opened
2024-07-09 01:59:20 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-09 01:59:20 [scrapy.extensions.telnet] INFO: Telnet console listening on 127.0.0.1:6023
2024-07-09 02:00:20 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-09 02:01:20 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-09 02:02:20 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2024-07-09 02:03:20 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
```

**Comment by elacuesta** [comment]
I was wrong about Conda, I was able to reproduce the issue in Windows 11 as well. What I found is that the spider works correctly if only one of the schemas is set in the `DOWNLOAD_HANDLERS` settings, e.g.:
```python
"DOWNLOAD_HANDLERS": {
    "https": "scrapy_playwright.handler.ScrapyPlaywrightDownloadHandler",
    # "http": "scrapy_playwright.handler.ScrapyPlaywrightDownloadHandler",
},
```
Fortunately plaintext http is falling out of favor nowadays, so only enabling the handler for the `https` schema should be enough for most sites.
I do not know why this is happening, I will need to check in more depth.

**Comment by FuturaTino** [comment]
> I was wrong about Conda, I was able to reproduce the issue in Windows 11 as well. What I found is that the spider works correctly if only one of the schemas is set in the `DOWNLOAD_HANDLERS` settings, e.g.:
> 
> ```python
> "DOWNLOAD_HANDLERS": {
>     "https": "scrapy_playwright.handler.ScrapyPlaywrightDownloadHandler",
>     # "http": "scrapy_playwright.handler.ScrapyPlaywrightDownloadHandler",
> },
> ```
> 
> Fortunately plaintext http is rare nowadays, so only enabling the handler for the `https` schema should be enough for most sites. I do not know why this is happening, I will need to check in more depth.

Use conda and set single https schema ❌ 
Dont use conda and set single https schema ✅
Yes, you are right. I use my default `python.exe`(venv also works) and delete the http schema and the spider runs successfully.
The bug has been circumvented, but the cause remains unknown.

**Comment by elacuesta** [comment]
Fixed as of [v0.0.40](https://github.com/scrapy-plugins/scrapy-playwright/releases/tag/v0.0.40).

**Comment by Onaga08** [comment]
Can anyone clear another doubt I have after reading this thread, my scrapy-playwright spider often processes a lot of requests, before getting stuck unexpectedly on something randomly! This seems unlikely to be because of the DOWNLOAD_HANDLERS settings, right?

**Comment by kirby-jack** [comment]
I have a strange case where using only `"http": "scrapy_playwright.handler.ScrapyPlaywrightDownloadHandler",` resolved this bug. When I set only `http` the websites were returning this. 

`'Not authenticated or invalid authentication credentials. '
                 'Make sure to update your proxy address, proxy username and '
                 'port.',`

` 'response_headers': {b'Content-Length': [b'121'],
                      b'Content-Type': [b'text/plain; charset=utf-8'],
                      b'Date': [b'Wed, 09 Apr 2025 15:57:53 GMT'],
                      b'Proxy-Authenticate': [b'Basic realm="The proxy you a'
                                              b're connecting is not in your'
                                              b' list."']},`

My only guess is my proxies are `http://` proxies so I don't know if this misalignment would make a difference? Just guessing here. 

I may further test using `https` and setting my proxies in another way as I am currently doing this in my middleware
```            request.meta["playwright_context_kwargs"] = {
                "java_script_enabled": True,
                "ignore_https_errors": True,
                "proxy": {
                    "server": server,
                    "username": username,
                    "password": password
                    }
                }```

**Comment by kirby-jack** [comment]
Can we please re-open this issue? 

I'm wondering based on the source code if it is getting stuck on the `await download_ready.wait()` line:

``` python annotate
        async def _handle_download(dwnld: Download) -> None:
            self.stats.inc_value("playwright/download_count")
            try:
                if failure := await dwnld.failure():
                    raise RuntimeError(f"Failed to download {dwnld.url}: {failure}")
                download_path = await dwnld.path()
                download["bytes"] = download_path.read_bytes()
                download["url"] = dwnld.url
                download["suggested_filename"] = dwnld.suggested_filename
            except Exception as ex:
                download["exception"] = ex
            finally:
                download_ready.set()

        page_goto_kwargs = request.meta.get("playwright_page_goto_kwargs") or {}
        page_goto_kwargs.pop("url", None)
        page.on("download", _handle_download)
        try:
            response = await page.goto(url=request.url, **page_goto_kwargs)
        except PlaywrightError as err:
            if not (
                self.config.browser_type_name in ("firefox", "webkit")
                and "Download is starting" in err.message
                or self.config.browser_type_name == "chromium"
                and "net::ERR_ABORTED" in err.message
            ):
                raise
            await download_ready.wait()
        finally:
            page.remove_listener("download", _handle_download)
```

This wait method will:
```python annotate
       """Block until the internal flag is true.

        If the internal flag is true on entry, return True
        immediately.  Otherwise, block until another task calls
        set() to set the flag to true, then return True.
        """
```

What if these stack up at the end of a spiders crawl and no new requests are being made? This will cause the spider to hang indefinitely.

**Comment by elacuesta** [comment]
The snippet you posted is outdated, there's additional code and logs in that part since https://github.com/scrapy-plugins/scrapy-playwright/pull/292, released as [v0.0.38](https://github.com/scrapy-plugins/scrapy-playwright/releases/tag/v0.0.38).
Granted, I could add some DEBUG log messages where the asyncio events are set, but I'd suggest first updating to the latest version to see if you're getting stuck on a `Waiting on download to finish for <URL>` message.

**Comment by Onaga08** [comment]
Still am facing this issue as well. Have deployed a work around to scrape in batches, but the underlying issue remains uncaught. 
My versioning looks like this 

`playwright = "^1.47.0"
scrapy = "^2.8.0"
scrapy-playwright = "0.0.42"`


Tried commenting the http download handler, my errback and context closing logic is sound, because I can see the headful browser close and open correctly, I haven't observed what happens with the browser and behaviour when it hangs because it hangs randomly and takes longer when observed! 

@kirby-jack you're not alone!
